content/IA-update_connections/connection-state-and-recovery.textile

```diff
@@ -0,0 +1,292 @@
+  [realtime.connection off:ARTRealtimeConnectionEventConnected listener:listener];
+
+  // remove a listener registered for all events
+  [realtime.connection off:listener];
+
+  // remove all event listeners
+  [realtime.connection off];
+```
+
+```[swift]
+  // remove a listener registered for a single event
+  realtime.connection.off(.connected, listener: listener)
+
+  // remove a listener registered for all events
+  realtime.connection.off(listener)
+
+  // remove all event listeners
+  realtime.connection.off()
+```
+
+h3(#handling-failures). Handling failures
+
+The client libraries will attempt to automatically recover from non-fatal error conditions. However, it will emit events to say what it's doing, so you can handle them yourself if you prefer.
+
+h4(#fatal-errors). Fatal errors
+
+Some classes of errors are fatal. These cause the connection to move to the @FAILED@ state. The client library will not attempt any automatic recovery actions. For example, if your token expires and the client library has no way to get a new token (so no authUrl and authCallback), the connection will enter the @FAILED@ state
+
+While the library will not automatically attempt to reconnect in the @FAILED@ state, explicit calls to "@connect()@":/api/realtime-sdk/connection#connect will make the client try again.
+
+h4(#nonfatal-errors). Nonfatal errors
+
+Other classes of error are nonfatal. For example, a client may have network connectivity issues. The library will attempt to automatically reconnect and recover from these sort of issues, as detailed in the @DISCONNECTED@ and @SUSPENDED@ explanations in the "Available connection states":#connection-states section.
+
+If message continuity is lost in the process, e.g. because you have been disconnected from Ably for more than two minutes, the library will notify you though the @resumed@-flag mechanism, detailed in the "Channels and Messages":/realtime/channels#nonfatal-errors page.
+
+h3(#connection-state-recovery). Connection state recovery
+
+p(caveat box). There are limitations to connection recovery. Once the client has been disconnected for more than 2 minutes, the client library moves into the suspended state indicating that all connection state is lost.  As such, all channels are automatically suspended at this point too indicating that channel continuity is not possible.  Once the connection is reestablished, the client library will reattach the suspended channels automatically and emit an attached event with the resumed flag set to false.  This ensures that as a developer, you can listen for attached events and check the resumed flag to see if a channel resumed fully and no messages were lost, i.e. when resumed is true, or the channel attached but could not resume, i.e. when resumed is false.  If channel continuity is not possible, then typically if historical messages are important to you, you would use the history API to retrieve all older messages, with the until the Attach flag set to true.
```

Comment:
I think is not the correct marker text for a caveat box (see Miro notes for correct format).

Reply:
Done.